### ISSUE TITLE
Fixed incompatibility of private property in PS 1.7.8.0

### ISSUE THREAD

**Issue by AliShareei** [PrestaShop/PrestaShop#26606]
### Prerequisites

- [x] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

Before PS 1.7.8.0 property $smarty ( in classes/pdf/PDF.php) was not declared in the class which is considered as public property.

In this version, apparently, an attempt has been made to clean the code and due to uncertainty, the data type has changed from public to private, which has broken the compatibility.

This problem is clearly seen in the modules that work with the PDF class, and the incompatibility is solved only by changing the type to a state that is available to other classes.

### Expected behavior

$smarty property should be declared as "public" or at least "protected"

### Steps to reproduce

1. create a module with override classes/pdf/PDF.php 
2. try to work with $smarty property on new class

### PrestaShop version(s) where the bug happened

1.7.8.0

### PHP version(s) where the bug happened

7.4

### If your bug is related to a module, specify its name and its version

custom module

**Comment by florine2623** [comment]
Hello @AliShareei ,

Could you send the module example so that we can test it out ?
Thanks!

**Comment by AliShareei** [comment]
> 
> 
> Hello @AliShareei ,
> 
> Could you send the module example so that we can test it out ? Thanks!

Hi
I think my explanation is simple enough.
Unfortunately, my module is complex but you can see the example code I write here.
This module work on all PrestaShop versions except 1.7.8.0 because of the problem I describe in the first post on this issue.

```
<?php
/**
 * Override for "classes/pdf/PDF.php"
 *
 * located in "modules/example_module/override/classes/pdf/PDF.php"
 */
 
class PDF extends PDFCore
{
    public function __construct($objects, $template, $smarty, $orientation = 'P')
    {
		parent::__construct($objects, $template, $smarty, $orientation);
		
		$this->smarty->assign(
			array(
				'test_var'	=> 'test_value',
			)
		);
		
    }
}
```

**Comment by florine2623** [comment]
I'll check with a develop to have more insight on this topic :)
Thanks for the details !

**Comment by matks** [comment]
Analyzis

1. PR https://github.com/PrestaShop/PrestaShop/pull/18220 introduced the private parameter `$smarty` as the property was being used before, but not declared.

2. This is not a bug, because `$smarty` was not declared before so it cannot be considered a part of our public API and consequently is not in our backward compatibility promise.

I know it worked before and it does not work with 1.7.8.0 but it's rather because you used a flawed property in the class. Its the famous usecase "it's a bug not a feature" 😄 .

3. I'm not sure we need to expose `$smarty`. The constructor allows you to inject `$smarty` so maybe rather than modifying `$smarty` directly, you can modify the `$smarty` instance and then inject it.

```
class PDF extends PDFCore
{
    public function __construct($objects, $template, $smarty, $orientation = 'P')
    {
                 $smarty->assign(
			array(
				'test_var'	=> 'test_value',
			)
		);
    
		parent::__construct($objects, $template, $smarty, $orientation);
    }
}
```

This works with 1.7.8.0 if I am not wrong.

4. That being said I think it's OK to make `$smarty` protected 👍  cannot hurt.

So to make it short
- it's not a bug, it's an improvement request
- OK to make it protected
- I think there is a workaround for 1.7.8.0

**Comment by AliShareei** [comment]
Hi @matks,

Thanks for your good review.
But you know in the real world we just stick with the sources code especially if a module is too old and we just update every year.

The property by default is `public` so in this case some change happened and make it incompatible.
in my case, `protected` will be enough. but without it maybe we have some issues that force us to rewrite some codes in extender class that in long term make more problems because every single change in PS core makes it incompatible.

I mean we can't do everything in the constructor like in my example code especially if we want to change something after core class in extender class so in my opinion changing the visibility of this property is necessary.

Thanks.

**Comment by matks** [comment]
> The property by default is public so in this case some change happened and make it incompatible.

Variable `$smarty` was initially not declared, which was the real bug.
As PHP is very tolerant, PHP accepts undeclared variable. This will maybe change in PHP 8.2 https://wiki.php.net/rfc/deprecate_dynamic_properties so as you can see, using an undeclared variable is really not safe.

Undeclared properties a.k.a dynamic properties are available as public by default so you used it. I'm sorry but you used something that was not supposed to exist ("it's not a bug it's a feature").

We could say a fair comparison is : imagine tomorrow someone in PrestaShop builds a /login API route and forgets to protect it Then payment gateway providers start using this route in their workflow, then the PrestaShop maintainer team understands the issue and protects the unsecure /login API route, but then every payment gateway providers who has built a module that use the /login API is failing. See the issue? An initial mistake was providing an edge usecase that was useful, and then when we fix the mistake this makes it not possible to exploit anymore.

**Comment by Progi1984** [comment]
Fixed by @matks in #28052 (and not tested).
